pkg/chain/ethereum/ethutil/adapter.go

```diff
@@ -0,0 +1,117 @@
+package ethutil
```

Comment:
We should not expect any breaking changes in `go-ethereum/common` and `go-ethereum/accounts/bind` in`v1.*.*`. Same for `celo-blockchain/common` and `celo-blockchain/accounts/bind`, especially that they are merging upstream periodically. 

I would consider keeping the structure and naming of interfaces in `ethtutil` 1:1 with `go-ethereum`/`celo-blockchain` representatives. This would let us navigate through the code easier. Also, adding new fields to header,  block, or transaction receipt structures would be a simple change in case we find it necessary in the future.

I have noted in the comments for `mining_waiter.go`, `nonce_manager.go`, and `block_counter.go` how such interfaces could look like. I would also recommend adding an interface in `ethutil.go` grouping them together. We can then document that this interface is expected to be implemented by all eth-like chain adapters:

```
type EthlikeChain interface {
  ChainReader
  TransactionReader
  ContractTransactor
}
```

Reply:
The problem with the approach you described it's the fact it will require a lot of boilerplate code. Types like `ethclient.Client` and `celoclient.Client` won't implement the `EthlikeChain` by default because of the concrete types like `common.Address`. We'll need to implement our own types and wrap the `go-ethereum/celo-blockchain` ones. Some converters will be also required. Are we sure we want to go that way? I doubt it will let us navigate through the code easier.

Reply:
I am not sure if we understand each other well here. `ChainReader`, `TransactionReader`, `ContractTransactor` are not types from `go-ethereum/` or `celo-blockchain/`. Those are types we declare and have the same names as their chain implementation counterparts - see earlier comments. They map only the subset of API but they map this subset 1:1 with their own types redeclared in `ethutil`. This is very similar to the approach you proposed except that the naming and interfaces are more aligned with what's in `go-ethereum/` or `celo-blockchain/`.

Reply:
Yep, I know. I meant the adapters code will be more complicated because we will have to implement interfaces like `ethlike.Header`, `ethlike.Block` twice (for Ethereum and for Celo). However, I see you also proposed to make those interfaces concrete types instead. I think this approach could work well. I will do the appropriate refactoring.

Reply:
Done. Please see:
- https://github.com/keep-network/keep-common/pull/64/commits/032237068f37d3c827f9caa29ff07b130fc554e8
- https://github.com/keep-network/keep-common/pull/64/commits/a8c0a5ca0a5d094451e6d66d2d5d4c3130c77bce
- https://github.com/keep-network/keep-common/pull/64/commits/b8472bbfbd85601516aca544a387fa8ec291d27d